### PR TITLE
Don't rewrite `n = m - c` if `m - n` overflows

### PR DESCRIPTION
**Pull request by lukemaurer**
The identity `n = m - c <=> c = m - n` only holds if `n`, `m`, and `c`
are all the same width. There are corner cases when `n` and `m` get
sign-extended and one of them is negative, most obviously when `n =
min_int` since then `-n = n`. For the moment, we have z3 telling us
that it suffices that `n` and `m` are both non-negative, so this
should keep us safe.

### CODE REVIEW

**Comment by lukemaurer** [comment]
Marking as draft while we make sure we got the z3 right.

**Comment by lthls** [comment]
Wouldn't it be simpler to use `Misc.no_overflow_sub m n` as condition ?

**Comment by xclerc** [comment]
Z3 script updated to use `Misc.no_overflow_sub` (same result as with `m` and `n` both positive):

```
(define-sort int63 () (_ BitVec 63))
(define-sort int64 () (_ BitVec 64))

(define-const z63 int63 ((_ int2bv 63) 0))

(declare-const m int63)
(declare-const n int63)
(declare-const c int64)

; let no_overflow_sub a b = (a lxor (lnot b)) lor (b lxor (a-b)) < 0
(define-fun no_overflow_sub ((a int63) (b int63)) Bool
  (bvslt (bvor (bvxor a (bvnot b)) (bvxor b (bvsub a b))) z63))

(assert (no_overflow_sub m n))

(assert (not (=
  (= ((_ sign_extend 1) n) (bvsub ((_ sign_extend 1) m) c))
  (= c ((_ sign_extend 1) (bvsub m n)))
)))

(check-sat)
```

**Comment by stedolan** [comment]
I don't understand this at all. Where is the sign-extension happening? This looks like a transformation of Cmm, where everything is a machine-word-sized twos-complement integer, so how can two values not be the same width?

I also don't understand the min_int example. It's true that `-n = n` when `n` is `min_int`, but that doesn't break the identity: `min_int = c - m` iff `c = m - min_int` iff `c = m + min_int` iff `c` and `m` differ in only the high bit.

**Comment by lthls** [comment]
The problem occurs because `Cconst_int` uses the type `int` to represent machine integers. So if we simplify `min_int - 1` at type int, then we get `max_int`, which is wrong as the operation should have been done at the target width (i.e. on 32 bits we should have gotten `0xc0000000`, a negative number, but get `0x40000000` instead).
A definitive fix could be to remove the `Cconst_int` constructor and use `Cconst_nativeint` everywhere (with `Targetint.t` insteaad of `Nativeint.t`). But this PR (if updated to use `Misc.no_overflow_sub`) merely follows the patterns of the functions in `Cmm_helpers` that simplify constant operations.

**Comment by stedolan** [comment]
Thanks for the explanation!

File: middle_end/flambda2/to_cmm/un_cps_helper.ml
Comment (lthls) [inline middle_end/flambda2/to_cmm/un_cps_helper.ml:177]: Nitpick: they don't get sign-extended at runtime, they get sign-extended later in the backend (during `Selection`). The problem is the same though.

**Review by lthls — COMMENTED** [review]
It would be nice if you could find a way to link to `Cmm_helpers`, which is full of such optimisations with associated overflow checks.

I'm not approving yet to let you decide if you want to fix the comment; once you're satisfied, let me know and I'll approve.

File: middle_end/flambda2/to_cmm/un_cps_helper.ml
Comment (lukemaurer) [reply middle_end/flambda2/to_cmm/un_cps_helper.ml:177]: Ah, good point. Better to say they'll _have been_ sign-extended by then.

**Review by lthls — APPROVED** [review]
There is a conflict that needs to be fixed, but apart from that it's good for me.

**Review by Gbury — APPROVED** [review]

**Pull request merged**
